src/core_modules/capture-core/metaDataStoreLoaders/programs/loadPrograms.js

```diff
@@ -1,110 +1,171 @@
+                .flatMap(programStage => getDataElementOptionSets(programStage.programStageDataElements));
+
+            return [
+                ...dataElementOptionSets,
+                ...getTrackedEntityAttributeOptionSets(program.programTrackedEntityAttributes),
+            ];
+        };
+
+        return (programsOutline): Array<Object> =>
+            programsOutline
+                .flatMap(program => getProgramOptionSets(program));
+    })();
+
+    const getTrackedEntityAttributeIds = stalePrograms =>
+        pipe(
+            () => stalePrograms
+                .flatMap(program =>
+                    (program.programTrackedEntityAttributes || [])
                         .map(programAttribute => programAttribute.trackedEntityAttributeId)
-                        .filter(TEAId => TEAId);
-
-                return [...accAttributeIds, ...attributeIds];
-            }
-            return accAttributeIds;
-        }, [])
-        : [];
-}
-
-function getCategories(missingPrograms) {
-    return missingPrograms
-        ? missingPrograms.reduce((accCategories, program) => {
-            const programCategories = program.categoryCombo &&
-                program.categoryCombo.categories;
-            return programCategories ? [...accCategories, ...programCategories] : [];
-        }, []) : [];
-}
-
-async function loadProgramBatch(programBatch) {
-    const programIds = programBatch.map(program => program.id);
-    const { convertedData: programs } = await storePrograms(programIds);
-    await storeProgramRules(programIds);
-    await storeProgramRulesVariables(programIds);
-    await storeProgramIndicators(programIds);
-    return programs;
-}
-
-export async function loadPrograms() {
-    const programsOutline = await queryProgramsOutline();
-    const optionSetsOutline = getOptionSetsOutline(programsOutline);
-    const missingPrograms = await getMissingPrograms(programsOutline);
-
-    const programBatches = chunk(missingPrograms, 50);
-    const programGroups = await Promise.all(
+                        .filter(TEAId => TEAId),
+                ),
+            attributeIds => [...new Set(attributeIds).values()],
+        )();
+
+    const getCategories = stalePrograms =>
+        pipe(
+            () => stalePrograms
+                .flatMap(program =>
+                    ((program.categoryCombo &&
+                    program.categoryCombo.categories) || []),
+                ),
+            categories => [
+                ...new Map(
+                    categories.map(ic => [ic.id, ic]),
+                ).values(),
+            ],
+        )();
+
+    const getTrackedEntityTypes = stalePrograms =>
+        pipe(
+            () => stalePrograms
+                .reduce((acc, program) => {
+                    program.trackedEntityTypeId && acc.add(program.trackedEntityTypeId);
+                    return acc;
+                }, new Set()),
+            trackedEntityTypeIdSet => [...trackedEntityTypeIdSet.values()],
```

Comment:
I have the feeling both functions can use `deduplicateArray` if you change them a little. Its a good way to reuse that code instead of repeating the logic here.

Reply:
Looks like you dont have the capacity to work on this improvement right now.